src/v/resource_mgmt/memory_sampling.cc

```diff
@@ -0,0 +1,241 @@
+    // have crossed the 10% and 20% marks we log the allocation sites. We stop
+    // afterwards.
+
+    size_t first_log_limit = _first_log_limit_fraction
+                             * seastar::memory::stats().total_memory();
+    size_t second_log_limit = _second_log_limit_fraction
+                              * seastar::memory::stats().total_memory();
+    size_t next_log_limit = first_log_limit;
+
+    while (true) {
+        try {
+            co_await _low_watermark_cond.wait([&next_log_limit]() {
+                auto current_low_water_mark
+                  = resources::available_memory::local()
+                      .available_low_water_mark();
+
+                return current_low_water_mark <= next_log_limit;
+            });
+        } catch (const ss::broken_condition_variable&) {
+            co_return;
+        }
+
+        auto allocation_sites = ss::memory::sampled_memory_profile();
+        const size_t top_n = std::min(size_t(5), allocation_sites.size());
+        top_n_allocation_sites(allocation_sites, top_n);
+
+        vlog(
+          _logger.info,
+          "{} {}",
+          diagnostics_header(),
+          fmt::join(
+            allocation_sites.begin(), allocation_sites.begin() + top_n, "|"));
+
+        if (next_log_limit == first_log_limit) {
+            next_log_limit = second_log_limit;
+        } else {
+            co_return;
+        }
+    }
+}
+
+memory_sampling::memory_sampling(
+  ss::logger& logger, config::binding<bool> enabled)
+  : memory_sampling(logger, std::move(enabled), 0.2, 0.1) {}
+
+memory_sampling::memory_sampling(
+  ss::logger& logger,
+  config::binding<bool> enabled,
+  double first_log_limit_fraction,
+  double second_log_limit_fraction)
+  : _logger(logger)
+  , _enabled(std::move(enabled))
+  , _first_log_limit_fraction(first_log_limit_fraction)
+  , _second_log_limit_fraction(second_log_limit_fraction) {
+    _enabled.watch([this]() { on_enabled_change(); });
+}
+
+void memory_sampling::on_enabled_change() {
+    // We chose a sampling rate of ~3MB. From testing this has a very low
+    // overhead of something like ~1%. We could still get away with something
+    // smaller like 1MB and have acceptable overhead (~3%) but 3MB should be a
+    // safer default for the initial rollout.
+    const size_t sampling_rate = 3000037;
+
+    // Note no logging here as seastar already logs about this
+    if (_enabled()) {
+        if (ss::memory::get_heap_profiling_sample_rate() == sampling_rate) {
+            return;
+        }
+
+        ss::memory::set_heap_profiling_sampling_rate(sampling_rate);
+    } else {
+        if (ss::memory::get_heap_profiling_sample_rate() == 0) {
+            return;
+        }
+
+        ss::memory::set_heap_profiling_sampling_rate(0);
+    }
+}
+
+void memory_sampling::start() {
+    setup_additional_oom_diagnostics();
+
+    // start now if enabled
+    on_enabled_change();
```

Comment:
I'm wondering how much we miss due to not having it started earlier. Have you checked the "up-sized memory allocated" from the pov of the sampler vs actual, after start?

Reply:
Yeah, it's fairly close. 

The only thing we are really missing is config loading plus the sizes of the objects. Just compared with the allocates stats and it's like 20MB.

Reply:
I see that's fine: it's just important that get in ahead of things like the chunk cache which allocate a substantial amount of memory (like 10% of the total or so).